blog-post/blog-post.md

```diff
@@ -0,0 +1,382 @@
+Core Solidity prototype uses a syntax which is similar, but not identical, to 
+Classic Solidity. Since our initial concern is to develop the new language type 
+system and its semantics, its surface syntax would probably change in the near 
+future.
+
+### Generics and Type Classes
+
+Core Solidity will introduce two new exciting abstraction mechanisms: generics and 
+type classes.
+
+Generics enable parametric polymorphism through type parameters, which allows the  
+implementation of algorithms and data structures that operate uniformly across 
+all types. As an example, we could define a polymorphic `identity` function:
+
+``` 
+forall T . function identity(x : T) -> T {
+  return x;
+}
+```
+
+While generic functions are interesting, most interesting operations are not defined 
+at all types. Overloading allows the definition of code which can operate in distinct 
+ways at different types. Type classes are the standard way of combining overloading and 
+parametric polymorphism (generics) in a systematic manner. A type class definition 
+declares the class name, its arguments and member functions type signature. As an 
+example, let's consider the task of defining addition over different types:
+
+```
+forall T . class T : Sum {
+  function sum (x : T, y : T) -> T;
+```
+
+Implementations for member functions for different types are provided by instance 
+definitions. As an example, let's consider the implementation of `Sum` for 
+`word` type which, internally, uses Yul assembly to perform addition. 
+
+```
+instance word : Sum {
+  function sum (x : word, y : word) -> word {
+    let res : word ;
+    assembly {
+      res := add(x,y);
+    }
+    return res;
+  }
+}
+```
+
+As another example, consider an instance for polymorphic pairs.
+
+```
+forall T1 T2 . T1 : Sum, T2 : Sum => instance (T1,T2) : Sum {
+  function sum (p1 : (T1,T2), p2 : (T1,T2)) -> (T1, T2) {
+    match p1, p2 {
+      (x1,y1), (x2, y2) => 
+        return (Sum.sum(x1,x2), Sum.sum(y1,y2));
+    }
+  }
+}
+```
+The previous definition shows that, whenever we have instances of 
+class `Sum` for types `a` and `b`, then we can also use function 
+`Sum.sum` on pairs of such types. The reader must noticed that
+in order to access the pair components we use **pattern matching**, 
+another feature which will be part of the Core Solidity.
+
+### Algebraic data types and pattern matching 
+
+Algebraic Data Types provide a way of modeling data modeling using sum types 
+(disjoint unions) and product types (structural records). This enables the construction of 
+precise types where invalid states are **unrepresentable** by design, thereby enhancing program 
+correctness through the type system itself. As an example, consider the following type 
+definition for an auction state. 
+
+```
+data AuctionState =
+    NotStarted (word) 
+  | Active (word, address) 
+  | Ended (word, address) 
+  | Cancelled;
+
```

Comment:
Wrapping this definition like this makes `|` look just like some arbitrary syntax decoration (as in `match`) rather than the sum operator. Putting it on a single line would makes it much clearer IMO:

```suggestion
data AuctionState =
    NotStarted(word) | Active(word, address) | Ended(word, address) | Cancelled;
``` 

Also, you have some stray empty line here.